daemon/debuginfo.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/api/v1/models"
+	. "github.com/cilium/cilium/api/v1/server/restapi/daemon"
+	"github.com/cilium/cilium/api/v1/server/restapi/endpoint"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/version"
+
+	"github.com/go-openapi/runtime/middleware"
+	"os"
+	"os/exec"
+)
+
+type getDebugInfo struct {
+	daemon *Daemon
+}
+
+// NewGetDebugInfoHandler returns the debug info endpoint handler for the agent
+func NewGetDebugInfoHandler(d *Daemon) GetDebuginfoHandler {
+	return &getDebugInfo{daemon: d}
+}
+
+func (h *getDebugInfo) Handle(params GetDebuginfoParams) middleware.Responder {
+	cver := version.GetCiliumVersion()
+	dr := models.DebugInfo{}
+	d := h.daemon
+
+	dr.CiliumVersion = fmt.Sprintf("%s (%s) %s", cver.Version, cver.Revision, cver.Arch)
+
+	kver, err := getKernelVersion()
+	dr.KernelVersion = fmt.Sprintf("%s", kver)
+
+	if err != nil {
+		dr.KernelVersion = fmt.Sprint("Error: %s\n", err)
+	}
+
+	status := getStatus(d)
+	dr.CiliumStatus = &status
+
+	var p endpoint.GetEndpointParams
+
+	dr.EndpointList = getEndpointList(p)
+	dr.Policy = getRulesList(h.daemon)
+
+	dr.CiliumMemoryMap = memoryMap(os.Getpid())
+
+	if d.nodeMonitor != nil && d.nodeMonitor.State() != nil {
+		dr.CiliumNodemonitorMemoryMap = memoryMap(d.nodeMonitor.GetPid())
+	}
+
+	return NewGetDebuginfoOK().WithPayload(&dr)
+}
+
+func memoryMap(pid int) string {
+	m, err := exec.Command("cat", fmt.Sprintf("/proc/%d/maps", pid)).CombinedOutput()
+	if err != nil {
+		return ""
+	}
+	return string(m)
+}
+
+func getRulesList(d *Daemon) *models.Policy {
```

Comment:
Let's move this function into `pkg/policy/repository.go`. That will make it easier to make the repository mutex private at a later point.